src/data/sign.lean

```diff
@@ -0,0 +1,200 @@
+def fin3_equiv : sign_type ≃* fin 3 :=
+{ to_fun :=  λ a, a.rec_on 0 (-1) 1,
+  inv_fun := λ a, match a with
+    | ⟨0, h⟩   := 0
+    | ⟨1, h⟩   := 1
+    | ⟨2, h⟩   := -1
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  left_inv :=  λ a, by cases a; refl,
+  right_inv := λ a, match a with
+    | ⟨0, h⟩   := rfl
+    | ⟨1, h⟩   := rfl
+    | ⟨2, h⟩   := rfl
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  map_mul' := λ x y, by casesm* _; refl }
+
+/-- Turn a `sign_type` into zero, one, or minus one. This is a coercion instance, but note it is
+only a `has_coe_t` instance: see note [use has_coe_t]. -/
+def cast {α} [has_zero α] [has_one α] [has_neg α] : sign_type → α
+| zero :=  0
+| pos  :=  1
+| neg  := -1
+
+@[simp] lemma cast_zero    {α} [has_zero α] [has_one α] [has_neg α] : cast 0 = (0 : α) := rfl
+@[simp] lemma cast_one     {α} [has_zero α] [has_one α] [has_neg α] : cast 1 = (1 : α) := rfl
+@[simp] lemma cast_neg_one {α} [has_zero α] [has_one α] [has_neg α] : cast (-1) = (-1 : α) := rfl
+
+instance {α} [has_zero α] [has_one α] [has_neg α] : has_coe_t sign_type α := ⟨cast⟩
+
+/-- `sign_type.cast` as a `mul_with_zero_hom`. -/
+@[simps] def cast_hom {α} [mul_zero_one_class α] [has_distrib_neg α] : sign_type →*₀ α :=
+{ to_fun    := cast,
+  map_zero' := rfl,
+  map_one'  := rfl,
+  map_mul'  := λ x y, by cases x; cases y; simp }
+
+end sign_type
+
+variables {α : Type*}
+
+open sign_type
+
+section preorder
+
+variables [has_zero α] [preorder α] [decidable_rel ((<) : α → α → Prop)] {a : α}
+
+/-- The sign of an element is 1 if it's positive, -1 if negative, 0 otherwise. -/
+def sign : α →o sign_type :=
+⟨λ a, if 0 < a then 1 else if a < 0 then -1 else 0, λ a b h, begin
+  dsimp,
+  split_ifs; try {constructor},
+  { cases lt_irrefl 0 (h_1.trans $ h.trans_lt h_3) },
+  { cases h_2 (h_1.trans_le h) },
+  { cases h_2 (h.trans_lt h_4) }
+  end⟩
+
+lemma sign_apply : sign a = ite (0 < a) 1 (ite (a < 0) (-1) 0) := rfl
+
+@[simp] lemma sign_zero : sign (0 : α) = 0 := by simp [sign_apply]
+@[simp] lemma sign_pos (ha : 0 < a) :  sign a = 1 := by rwa [sign_apply, if_pos]
+@[simp] lemma sign_neg (ha : a < 0) : sign a = -1 := by rwa [sign_apply, if_neg $ asymm ha, if_pos]
+
+end preorder
+
+section linear_order
+
+variables [has_zero α] [linear_order α] {a : α}
+
+lemma sign_ne_zero (h : a ≠ 0) : sign a ≠ 0 :=
+begin
+  contrapose! h,
+  rw sign_apply at h,
+  split_ifs at h,
+  { cases h },
+  { cases h },
+  exact ((lt_trichotomy a 0).resolve_left h_2).resolve_right h_1
+end
+
+end linear_order
+
+section linear_ordered_ring
+
+variables [linear_ordered_ring α] {a b : α}
+
+/- I'm not sure why this is necessary, see https://leanprover.zulipchat.com/#narrow/stream/
+113488-general/topic/type.20class.20inference.20issues/near/276937942 -/
+local attribute [instance] linear_ordered_ring.decidable_lt
+
+-- TODO: determine whether `ring` is required. Current thinking is that it is, for the implication
+-- a < 0 → b < 0 → 0 < a * b.
```

Comment:
I am not sure how far you want to take this, but if you take the boolean algebra `set unit`, using `set.univ` as `0` and `sup` as multiplication, then `sign (∅ ⊔ ∅) ≠ sign ∅ * sign ∅`, since `sign ∅` is `-1` and `∅ ⊔ ∅ = ∅`.

I think that this is another way of saying that you take the field `zmod 2`, and you define `1 < 0`.

It may be that requiring `0 ≤ 1` is sufficient, once you have a monoid, but I am not sure.

Reply:
0 <= 1 is part of the axioms on an ordered semiring, sadly. There's been some discussion on this on the discord, but no conclusive answer yet

Reply:
(another similar example that breaks because of ¬0 ≤ 1 is order_dual \N)

Reply:
If you want an example of a ring with a linear order (but of course not a `linear_ordered_ring`) in which `0 < 1` and containing two negative elements whose product is negative, the code below appears to work!

```lean
import data.zmod.basic

namespace Z8
def fff : zmod 8 → ℕ := λ x, 
begin
  rcases x with ⟨_ | _ | _ | _ | _ | _ | _ | _ | x, hx⟩,
  exacts [3+ 0, 3+ 1, 3 - 1, 3 + 3, 3 - 3, 3 + 5, 3 - 2, 3 + 7, 0]
end

instance : linear_order (zmod 8) :=
linear_order.lift fff $ by dec_trivial

lemma not_si : ((0 : zmod 8) < 1) ∧ ∃ a b : zmod 8, (a < 0) ∧ (b < 0) ∧ (a * b < 0) :=
dec_trivial

end Z8
```

The unique field of `linear_ordered_ring` that fails is `add_le_add_left`.

Reply:
So the conclusion is that indeed, `linear_ordered_ring` is a necessary condition?
```suggestion
```